### DISCUSSION TITLE
Best approach for returning derived/owned data from `session_auth_hash`?

### DISCUSSION THREAD

**Discussion by sbking** [maxcountryman/axum-login#209] [Q&A]
Rather than duplicating actual password hashes (which currently only are stored in the Credentials table) into my Users and Sessions table, I want to derive the `auth_hash` from properties on a User such as `password_changed_at` and `sessions_invalidated_at`. This way I should be able to achieve the same effect of invalidating sessions when a user changes their password, without copying my actual Argon2 hashes all over the place.

However the `AuthUser::session_auth_hash` must return borrowed data, which makes it difficult to return a derived value. It seems like I must have an `auth_hash` field owned by my `User` struct, which complicates other things like constructing, serializing, and deserializing users. It would be nice if `session_auth_hash` could return a `Cow` or something. Or is there another approach I'm missing?

**Comment by maxcountryman** [comment]
`Cow` could be a good direction to consider, I'd need to think more on that.

> However the AuthUser::session_auth_hash must return borrowed data, which makes it difficult to return a derived value.

The problem here is that you're deriving from borrowed fields on your user struct, right? Your use case makes a lot of sense to me so I'd like to think on how we can better support it.

**Reply by sbking** [reply]
The `password_changed_at` and `sessions_invalidated_at` are owned fields on my `User` struct. Basically what I'm trying to do is concatenate these two unix timestamps and hash them with SHA3 or whatever, and then return that digest. But this process requires _something_ to own the SHA digest, and I'd rather not make my `User` struct own it.

**Reply by maxcountryman** [reply]
Got it, that makes sense.

Could you return a Vec from a user method instead?

I'm not opposed to making the interface more flexible but it's also a breaking change so I do want to make sure we think through the design space as thoroughly as we can.

**Reply by sbking** [reply]
Hmm, what would that look like? Wouldn't that have the same issue, where the Vec has to have an owner, and the borrow checker will complain that `session_auth_hash` returns a borrow of the Vec?

Another option I'm thinking about is just defining a wrapper struct like:

```rs
pub struct AuthUser {
    pub user: User,
    pub auth_hash: [u8; 32],
}

impl axum_login::AuthUser for AuthUser {
    // ..
}
```

Which doesn't feel great but it should work. Otherwise yeah I'm not sure how to do something better without a breaking change. Maybe a feature flag or a new `AuthUserDerived`/`AuthUserCow` (naming things) trait that has a blanket implementation over `AuthUser`?

**Reply by maxcountryman** [reply]
You're right, without storing the value on the user struct (or some wrapper, etc) then nothing owns the derived hash.

Cow or Vec seem like better options and I'm open to making a breaking change.

I also wanted to point out an important security consideration about your specific implementation: deriving the hash solely from timestamps could be dangerous as it's far easier to guess a timestamp than it would be e.g. a cryptographically securely generated password hash. So what this means is that I would recommend including more than those two fields and ideally some kind of value derived from a cryptographic key (sha256 should help, but it would be better to use with a salt or a different hash altogether, like argon2). All this to say if you have the password hash then using it in some way for this seems like a more secure approach (I understand it's on another table, so I realize the logistics are more challenging in this case).

**Reply by sbking** [reply]
Hmm, can you help me understand the security risk there? I could use an HMAC and salt or something like Argon2, but even if someone were to guess the timestamps how would they actually forge a session and insert it into the database? I thought the session hash was mainly a way to invalidate sessions without having to batch delete them.

**Reply by maxcountryman** [reply]
> but even if someone were to guess the timestamps how would they actually forge a session and insert it into the database

The attack vector assumes some kind of access to the session (not necessarily the database). While it might seem unlikely, as a defense-in-depth measure, this is mitigated by ensuring the data of the auth hash is itself cryptographically secure and practically unguessable

It's noteworthy that [the Django module](https://docs.djangoproject.com/en/5.0/topics/auth/default/#using-the-django-authentication-system) this crate is loosely based on provides a default implementation which uses HMAC over the password hash and I would suggest following a similar pattern unless there's a good (and clearly secure) reason not to.

**Reply by sbking** [reply]
Thinking about it more, instead of copying Argon2 password hashes to multiple tables, I'd rather just store something like a 256-bit CSPRNG random token on the User table, and regenerate the token when the user changes their password or a security event occurs. In which case I would be able to just return a borrow slice.

Although I'd still like to understand better why the hash needs to be unguessable from a security perspective - it feels like for an attacker to forge a session they would already need to have write access to my database as well as the HMAC key used to hash session IDs before storing sessions in the database, in which case the entire app is already compromised.

**Reply by sbking** [reply]
I think Django's default implementation works well precisely because it derives a hash from the actual password hash stored on the user object, rather than copying the actual password hash around - that way an application developer doesn't have to think about generating a new session validation token whenever a user changes their password. So I do think there is still a strong argument for the return type to be something like Cow or Vec

**Reply by maxcountryman** [reply]
>  I'd rather just store something like a 256-bit CSPRNG random token on the User table, and regenerate the token when the user changes their password or a security event occurs

This is a pattern I've seen in mature web applications; similarly useful for password resets, etc.

> it feels like for an attacker to forge a session they would already need to have write access to my database

The session, yes; but I'm not following why the database would be required.

Generally speaking, the session is a weaker vector which is why e.g. things like session fixation are a problem.

**Reply by sbking** [reply]
> The session, yes; but I'm not following why the database would be required.

In my case:

- The session cookie is just a random 128bit identifier, signed with a secret key
- The session database table doesn't store the actual session ID - the primary key is an HMAC of the session ID using another secret key
- The session data, including the session auth hash, is only stored in the database and never exposed to the user

That's why I'm not quite understanding the attack vector or how the session auth hash prevents session fixation or session forging or why it actually needs to be unguessable. I'm not sure how an attacker can use a known session auth hash to exploit any attack vector that doesn't involve them already having write access to the database and access to at least one secret key.

**Reply by maxcountryman** [reply]
Consider that an application may be vulnerable to inadvertent session data insertion, as one example.

If such a vulnerability is found, then the ability to guess an auth hash is suddenly a serious security problem.

In my experience it's not a vector in isolation, but instead of series of failures, that leads to a compromise. For that reason, there's a culture of biasing toward defense-in-depth tactics when it comes to web application security.

**Reply by sbking** [reply]
Okay, I think that makes sense. In Django's case I think it makes a little more sense as I've seen more than one production Django app using signed (unencrypted) cookie sessions, where an attacker would only need to know the cookie signing secret to insert a fake session auth hash. All-in-all I think I will go with a random token on the user instead of one that is derived from the password hash, but I'm looking forward to any update that would allow us to do something more like the Django implementation and use an HMAC that's not actually stored on the user object!

**Reply by maxcountryman** [reply]
> I'm looking forward to any update that would allow us to do something more like the Django implementation and use an HMAC of the actual password hash!

I agree. This is the right direction from my view. There's room to make breaking changes, especially prior to the 1.0 release.
